contracts/service-provider-directory/src/contract.rs

```diff
@@ -0,0 +1,330 @@
+use crate::{
+    error::ContractError,
+    msg::{ConfigResponse, ExecuteMsg, InstantiateMsg, QueryMsg, ServicesListResponse},
+    state::{Config, Service, CONFIG, SERVICES},
+};
+use cosmwasm_std::{
+    to_binary, Addr, Binary, Deps, DepsMut, Env, MessageInfo, Order, Response, StdResult,
+};
+use cw2::set_contract_version;
+
+// version info for migration info
+const CONTRACT_NAME: &str = "crate:nym-service-provider-directory";
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+pub fn instantiate(
+    deps: DepsMut<'_>,
+    _env: Env,
+    info: MessageInfo,
+    msg: InstantiateMsg,
+) -> StdResult<Response> {
+    let config = Config {
+        updater_role: msg.updater_role.clone(),
+        admin: msg.admin.clone(),
+        deposit_required: msg.deposit_required.clone(),
+    };
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+    CONFIG.save(deps.storage, &config)?;
+
+    Ok(Response::new()
+        .add_attribute("method", "instantiate")
+        .add_attribute("owner", info.sender)
+        .add_attribute("admin", msg.admin)
+        .add_attribute("updater_role", msg.updater_role))
+}
+
+pub fn execute(
+    deps: DepsMut<'_>,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::Announce {
+            nym_address: client_address,
+            service_type,
+            owner,
+        } => execute::announce(deps, env, info, client_address, service_type, owner),
+        ExecuteMsg::Delete { service_id: sp_id } => execute::delete(deps, info, sp_id),
+    }
+}
+
+pub mod execute {
+    use cosmwasm_std::{coins, BankMsg, Coin};
+
+    use super::*;
+    use crate::state::{self, NymAddress, ServiceId, ServiceType};
+
+    /// Announce a new service. It will be assigned a new service provider id.
+    pub fn announce(
+        deps: DepsMut,
+        env: Env,
+        info: MessageInfo,
+        nym_address: NymAddress,
+        service_type: ServiceType,
+        owner: Addr,
+    ) -> Result<Response, ContractError> {
+        let deposit_required = state::deposit_required(deps.storage)?;
+        let denom = deposit_required.denom.clone();
+        let will_deposit = cw_utils::must_pay(&info, &denom)
+            .map_err(|err| ContractError::DepositRequired { source: err })?;
+
+        if will_deposit < deposit_required.amount {
+            return Err(ContractError::InsufficientDeposit {
+                funds: will_deposit,
+                deposit_required,
+            });
+        }
+
+        if will_deposit > deposit_required.amount {
+            return Err(ContractError::TooLargeDeposit {
+                funds: will_deposit,
+                deposit_required,
+            });
+        }
+
+        let admin = state::admin(deps.storage)?;
+
+        let will_deposit = Coin::new(will_deposit.u128(), denom);
+        let deposit_msg = BankMsg::Send {
```

Comment:
We don't need to send those tokens to the admin. This would make things even more complicated when people decide to remove their service providers. Just leave those tokens at the contnract